walletclient.py

```diff
@@ -151,11 +245,115 @@ def check():
         for webcash, result in response["results"].items():
             if result["spent"] in (None, True):
                 print(f"Invalid webcash found: {str(webcash)}; removing from wallet")
-                if webcash not in batch:
+
+                # Use this as the key so that amount differences don't cause an
+                # item-not-found error on otherwise same webcash.
+                webcash_hashed_value = PublicWebcash.deserialize(webcash).hashed_value
+
+                if webcash_hashed_value not in batch:
                     raise Exception(f"Server-returned webcash {str(webcash)} wasn't in our request.  This should never happen!")
-                webcash_wallet["unconfirmed"].append(batch[webcash])
-                webcash_wallet["webcash"].remove(batch[webcash])
 
+                webcash_wallet["unconfirmed"].append(batch[webcash_hashed_value])
+                webcash_wallet["webcash"].remove(batch[webcash_hashed_value])
+
+    save_webcash_wallet(webcash_wallet)
+
+@cli.command("check")
+@lock_wallet
+def check():
+    return check_wallet()
+
+@cli.command("recover")
+@lock_wallet
+def recover():
+    """
+    Recover webcash from a webcash wallet using its master secret as a
+    deterministic seed. Also check all webcash in the wallet.
+    """
+
+    webcash_wallet = load_webcash_wallet()
+
+    # gaplimit is the maximum window span that will be used, on the assumption
+    # that any valid webcash will be found within the last item plus gaplimit
+    # number more of the secrets.
+    gaplimit = 20
+
+    # Check all the webcash in the wallet and remove any webcash that has been
+    # already spent.
+    check_wallet()
+
+    for chain_code in webcash_wallet["walletdepths"].keys():
+        # keep track of where we're at in this process
+        current_walletdepth = 0
+
+        reported_walletdepth = webcash_wallet["walletdepths"][chain_code]
+
+        # Iterate through gaplimit-many secrets at a time and check each one.
+        _idx = 0
+        last_used_walletdepth = 0
+        has_had_webcash = True
+        while has_had_webcash:
+            print(f"Checking gaplimit {gaplimit} secrets for chaincode {chain_code}, round {_idx}...")
+
+            # assume this is the last iteration
+            has_had_webcash = False
+
+            # Check the next gaplimit number of secrets. Continue to the next round
+            # if any of the secrets have ever been used, regardless of whether they
+            # still have value.
+
+            check_webcashes = {}
+            for x in range(current_walletdepth, current_walletdepth + gaplimit):
+                secret = generate_new_secret(webcash_wallet, chain_code=chain_code, walletdepth=x)
+                webcash = SecretWebcash.deserialize("e1:secret:" + secret)
```

Comment:
nit: why is this "e1:..." ? I thought 0 was the stand-in for "unknown amount"

Reply:
server returns an error for 0 because 0 is an invalid amount, this could be fixed if it's a problem tho

Reply:
Not a problem now. But maybe something to change in the future.